### ISSUE TITLE
twisted.conch.scripts.conch doesn't work on Python 3, b'bad host key', 9

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8791]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)| @rodrigc reported|
|-|-|
|Trac ID|trac#8791|
|Type|enhancement|
|Created|2016-08-27 17:33:11Z|
|Branch|https://github.com/twisted/twisted/tree/8791-rodrigc-conch|

If I do the following on Python 2.7 on my Mac:

```
python -m twisted.conch.scripts.conch
```

That displays the help for the conch client,
and if I do:

```
python -m twisted.conch.scripts.conch 192.168.1.7
```

then I can log into a machine with IP address 192.168.1.7.

On Python 3, I can try to do:

```
python -m twisted.conch.scripts.conch 192.168.1.7
```

then I get:

```
Connection to 127.0.0.1 closed.
conch: exiting with error [Failure instance: Traceback (failure with no frames): <class 'twisted.conch.error.ConchError'>: (b'bad host key', 9)
]
```

<details><summary>Searchable metadata</summary>

```
trac-id__8791 8791
type__enhancement enhancement
reporter__rodrigc rodrigc
priority__normal normal
milestone__Python_3_x Python-3.x
branch__8791_rodrigc_conch 8791-rodrigc-conch
branch_author__ 
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__None None
time__1472319191942332 1472319191942332
changetime__1600113263424915 1600113263424915
version__None None
owner__rodrigc rodrigc
cc__z3p
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_2"></a>|@rodrigc commented|
|-|-|

I did some analysis, and found the problem in twisted/conch/client/knownhosts.py:

```
305     def matchesHost(self, hostname):
306         """
307         Implement L{IKnownHostEntry.matchesHost} to compare the hash of the
308         input to the stored hash.
309 
310         @param hostname: A hostname or IP address literal to check against this
311             entry.
312         @type hostname: L{bytes}
313 
314         @return: C{True} if this entry is for the given hostname or IP address,
315             C{False} otherwise.
316         @rtype: L{bool}
317         """
318         return (_hmacedString(self._hostSalt, hostname) == self._hostHash)
319 
```

Line 318 throws an exception and never returns a value.  I changed the code slightly:

```
    try:
        h =  (_hmacedString(self._hostSalt, hostname) == self._hostHash)
    except Exception e:
        print("ERROR: ", str(e))

    return h
```

and found that the error was:

```
ERROR:  Unicode-objects must be encoded before hashing <class 'TypeError'>
```

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_3"></a>|@rodrigc commented|
|-|-|

https://github.com/twisted/twisted/pull/589

**Comment by twisted-trac** [comment]
|[<img alt="acabhishek942's avatar" src="https://avatars.githubusercontent.com/u/7778186?s=50" width="50" height="50">](https://github.com/acabhishek942)<a name="note_4"></a>|@acabhishek942 set owner to @rodrigc|
|-|-|

reviewed on github

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_6"></a>|@rodrigc set status to `closed`|
|-|-|

In changeset 9547189afbd4d0c47c609c9031f2322f39de8c5d
```
#!CommitTicketReference repository="" revision="9547189afbd4d0c47c609c9031f2322f39de8c5d"
Merge pull request #589 from twisted/8791-rodrigc-conch

Author: rodrigc
Reviewer: acabhishek942
Fixes: #8791

Fix conch and cftp under Python 3
```

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_7"></a>|@rodrigc commented|
|-|-|

In changeset e1c54d7e72efa10e578fe50211a4cdebc5d0ee3c
```
#!CommitTicketReference repository="" revision="e1c54d7e72efa10e578fe50211a4cdebc5d0ee3c"
Merge pull request #1383 from graingert/patch-9

Author: graingert
Reviewer: rodrigc
Fixes: ticket:8791

Delete 8791.feature
```
